### PR TITLE
Implement: Priorities

### PR DESCRIPTION
**Pull request by fran-cap**
Adds the priorities panel used to set ai tasks.
Created constants for inc/dec and max priorities values.
Changed data types of priority list type and name to match decompiliation (double check this).

Implements:
	LegoRR::Priorities_LoadImages
	LegoRR::Priorities_Reset
	LegoRR::Priorities_LoadLevel
	LegoRR::Priorities_LoadPositions
	LegoRR::Priorities_HandleInput
	LegoRR::Priorities_ChangePriorityPressed
	LegoRR::Priorities_SetHoverArea
	LegoRR::Priorities_MovePriorityUpOrTop
	LegoRR::Priorities_MovePriorityTop
	LegoRR::Priorities_MovePriorityUp
	LegoRR::Priorities_TurnPriorityOff
	LegoRR::Priorities_MoveCursorToPriorityUpButton
	LegoRR::Priorities_MoveCursorToPriorityImage
	LegoRR::Priorities_Draw

### CODE REVIEW

**Review by trigger-segfault — COMMENTED** [review]
Good work. Some changes need to be made, and there's one bug (which is actually my fault because it was like that in the comments I wrote).

File: src/openlrr/game/interface/Priorities.cpp [hunk 1]
```diff
@@ -28,45 +39,365 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
...
+	char* keyPath;
+	char* tempPositionValue;
+	int orderIndex;
+	char* parts[16];
+	char* startGameName = (char*)gameName;
+	orderIndex = 0;
+
+	for(uint32 i = 0; i < AI_Priority_Count; i++){
+		orderIndex++;
+		_itoa(orderIndex, (char*)&gameName, 10);
```
[the remark this answers is not in the document]
Comment (fran-cap) [reply src/openlrr/game/interface/Priorities.cpp:123]: Small note here, due to it being orderIndex++; we need to do a +1 on i. Also VS throws an error for itoa not being part of std so I kept ht _itoa:

```C++
char orderName[12]; // Max number of characters +1 that can appear in signed int32.
_itoa((int)i+1, orderName, 10);
```

**Comment by fran-cap** [comment]
Ok this should capture the above changes.
Thanks for verifying that I'm not crazy with that disable message ;).

File: src/openlrr/game/interface/Priorities.cpp [hunk 2]
```diff
@@ -28,45 +39,361 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
...
+/* <PriorityType> <LangName>[|LangSFX]:<NormalImage>:<Pressedmage>:<OffImage> */
+void __cdecl LegoRR::Priorities_LoadImages(const Gods98::Config* config, const char* gameName)
+{
+	AI_Priority priorityType;
+	char* stringParts[10];
+	char* stringPartsSub[10];
+	char buff[512];
+	
+	const char* arrayID = Config_ID(gameName, "PriorityImages");
+	for (const Gods98::Config* prop = Config_FindArray(config, arrayID); prop != nullptr;  prop = (Gods98::Config*)Config_GetNextItem(prop)) {
```
Comment (trigger-segfault) [inline src/openlrr/game/interface/Priorities.cpp:51]: `(Gods98::Config*)` cast needs to be removed for `Config_GetNextItem`.

File: src/openlrr/game/interface/Priorities.cpp [hunk 3]
```diff
@@ -28,45 +39,361 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
...
+}
 
 // <LegoRR.exe @0045d210>
-//bool32 __cdecl LegoRR::Priorities_LoadLevel(const Gods98::Config* config, const char* gameName, const char* levelName);
+bool32 __cdecl LegoRR::Priorities_LoadLevel(const Gods98::Config* config, const char* gameName, const char* levelName)
+{
+	char* arrayID;
+	const Gods98::Config* prop;
+	prioritiesGlobs.count = 0;
+	arrayID = (char*)Config_ID(gameName, levelName, "Priorities");
```
Comment (trigger-segfault) [inline src/openlrr/game/interface/Priorities.cpp:87]: `arrayID` needs to be const and `(char*)` cast should be removed.

File: src/openlrr/game/interface/Priorities.cpp [hunk 4]
```diff
@@ -28,45 +39,361 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
...
+void __cdecl LegoRR::Priorities_LoadPositions(const Gods98::Config* config, const char* gameName)
+{
+	char* keyPath;
+	char* parts[16];
+	char* startGameName = (char*)gameName;
+
+	for (uint32 i = 0; i < AI_Priority_Count; i++) {
+		char orderName[12]; // Max number of characters +1 that can appear in signed int32.
+		_itoa((int)i+1, orderName, 10);
+		keyPath = (char*)Config_ID(startGameName, "PrioritiesImagePositions", &orderName);
```
Comment (trigger-segfault) [inline src/openlrr/game/interface/Priorities.cpp:122]: `keyPath` needs to be const and `(char*)` cast should be removed.

`startGameName` isn't needed anymore, so pass `gameName` instead and remove the local variable.

File: src/openlrr/game/interface/Priorities.cpp [hunk 5]
```diff
@@ -28,45 +39,361 @@ LegoRR::Priorities_Globs & LegoRR::prioritiesGlobs = *(LegoRR::Priorities_Globs*
...
+	AI_Priority currPriorityType;
+
+	if (orderIndex > 0 && orderIndex < prioritiesGlobs.count) {
+		currPriorityType = prioritiesGlobs.buttonTypes[orderIndex];
+		for(index = orderIndex; index > 0; index--){
+				/* globs::prioritiesGlobs.priorityTypeTable_1[index - 1] */
+				//priorityType = (AI_Priority)prioritiesGlobs.sfxPriorityName[index + 0x1a];
+				priorityType = prioritiesGlobs.buttonTypes[index-1];
+				prioritiesGlobs.buttonTypes[index] = priorityType;
+				AITask_Game_SetPriorityValue(priorityType, MAX_PRIO + (index * -PRIO_INC));
```
Comment (trigger-segfault) [inline src/openlrr/game/interface/Priorities.cpp:242]: This area is one tab level too high.

**Review by trigger-segfault — COMMENTED** [review]
Heh yeah, I remember questioning myself with the isOff section back when I was commenting that in Ghidra. Good thing the result of that condition was easy to identify, or this would have gone unnoticed.

Good catch with the 1-indexed order index for itoa. :)

A few more changes. And the `orderIndex > 0` is still missing from MovePriorityUp.

**Comment by fran-cap** [comment]
Thanks for the 2nd review. Pushed up newest changes.

**Comment by trigger-segfault** [comment]
Everything looks good. Thanks for your work. :)

**Pull request merged**